pkg/job/scraper.go

```diff
@@ -0,0 +1,240 @@
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/nerdswords/yet-another-cloudwatch-exporter/pkg/job/cloudwatchrunner"
+
+	"github.com/nerdswords/yet-another-cloudwatch-exporter/pkg/clients/account"
+	"github.com/nerdswords/yet-another-cloudwatch-exporter/pkg/logging"
+	"github.com/nerdswords/yet-another-cloudwatch-exporter/pkg/model"
+)
+
+type Scraper struct {
+	jobsCfg       model.JobsConfig
+	logger        logging.Logger
+	runnerFactory runnerFactory
+}
+
+type runnerFactory interface {
+	GetAccountClient(region string, role model.Role) account.Client
+	NewResourceMetadataRunner(logger logging.Logger, region string, role model.Role) ResourceMetadataRunner
+	NewCloudWatchRunner(logger logging.Logger, region string, role model.Role, job cloudwatchrunner.Job) CloudwatchRunner
+}
+
+type ResourceMetadataRunner interface {
+	Run(ctx context.Context, region string, job model.DiscoveryJob) ([]*model.TaggedResource, error)
+}
+
+type CloudwatchRunner interface {
+	Run(ctx context.Context) ([]*model.CloudwatchData, error)
+}
+
+func NewScraper(logger logging.Logger,
+	jobsCfg model.JobsConfig,
+	runnerFactory runnerFactory,
+) *Scraper {
+	return &Scraper{
+		runnerFactory: runnerFactory,
+		logger:        logger,
+		jobsCfg:       jobsCfg,
+	}
+}
+
+func (s Scraper) Scrape(ctx context.Context) ([]model.TaggedResourceResult, []model.CloudwatchMetricResult, []Error) {
+	// Setup so we only do one GetAccount call per region + role combo when running jobs
+	roleRegionToAccount := map[model.Role]map[string]func() string{}
+	jobConfigVisitor(s.jobsCfg, func(_ any, role model.Role, region string) {
+		if _, exists := roleRegionToAccount[role]; !exists {
+			roleRegionToAccount[role] = map[string]func() string{}
+		}
+		roleRegionToAccount[role][region] = sync.OnceValue[string](func() string {
+			accountID, err := s.runnerFactory.GetAccountClient(region, role).GetAccount(ctx)
+			if err != nil {
+				s.logger.Error(err, "Failed to get Account", "region", region, "role_arn", role.RoleArn)
+				return ""
+			}
+			return accountID
+		})
+	})
+
+	var wg sync.WaitGroup
+	mux := &sync.Mutex{}
+	jobErrors := make([]Error, 0)
+	metricResults := make([]model.CloudwatchMetricResult, 0)
+	resourceResults := make([]model.TaggedResourceResult, 0)
+	s.logger.Debug("Starting job runs")
+
+	jobConfigVisitor(s.jobsCfg, func(job any, role model.Role, region string) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			var namespace string
+			jobAction(s.logger, job, func(job model.DiscoveryJob) {
+				namespace = job.Type
+			}, func(job model.CustomNamespaceJob) {
+				namespace = job.Namespace
+			})
+			jobLogger := s.logger.With("namespace", namespace, "region", region, "arn", role.RoleArn)
+
+			accountID := roleRegionToAccount[role][region]()
+			if accountID == "" {
+				jobError := Error{
+					AccountID: "",
+					Region:    region,
+					RoleARN:   role.RoleArn,
+					Namespace: namespace,
+					Message:   "Account for job was not found see previous errors",
```

Comment:
```suggestion
					Message:   "Account for job was not found. There should be a corresponding 'Failed to get Account' error logged previously.",
```

Just to hopefully tie it a bit more specifically so that operators will know what they're looking for.

Reply:
Alternatively perhaps the `roleRegionToAccount` mapped function could have a return signature of `(string, err)` and directly check for `if err != nil` and add the error in this struct?

Reply:
I like it!